distributed/scheduler.py

```diff
@@ -7225,19 +7402,29 @@ def adaptive_target(self, target_duration=None):
         target_duration = parse_timedelta(target_duration)
 
         # CPU
+
+        # FIXME maintain a proper estimate of queued occupancy!!
+        # This is merely a hack intended to make queuing sorta work with adaptive scaling
+        # so people can try it out in the short term (at least the cluster should scale up
+        # when tasks are queued).
+        avg_duration = (
+            (self.total_occupancy / self.total_nthreads) if self.total_nthreads else 0
+        )
+        queued_occupancy = len(self.queued) * avg_duration
```

Comment:
The way I see it, there are two use cases for total_occupancy

1. Work stealing. More specifically `Scheduler.check_idle_saturated`. It's not that obvious but the `saturated` set is actually **only** used for work stealing. idle is also strongly impacting work stealing but it is also used in decide_worker. Therefore, messing with total_occupancy can change the definition of both idle and saturated and will impact potentially impact a lot of things. 
That said, I've been suspicious about these definitions for a very long time but never had concrete examples showing that this is false. 

https://github.com/dask/distributed/blob/a1d2011e7d5d5ba26bea8bd4b0dec65f485f2b43/distributed/scheduler.py#L2600-L2616

2. adaptive scaling

I'm much less concerned about adaptive_target. We should definitely not *break* it and use _sane_ heuristics but if this definition shifts a bit, I think we're good, particularly if the initial behavior with toggle off is not changing (that's the case if I read correctly)

> I'm hoping this actually eliminates most of the need for work stealing overall.

It likely reduces importance and likely reduces complexity but I doubt we can remove it entirely. I suggest to not work under the assumption that we can throw everything away after this. Particularly, there are also work loads which are not root task dominated that still require some rebalancing.
See also https://github.com/dask/distributed/issues/6600 for a conversation about use cases for work stealing

Reply:
I think the new definition of idle to incorporate whether there are free slots makes a lot of sense.

> I didn't change the saturated definition because I don't honestly understand what it's trying to measure: https://github.com/dask/distributed/blob/a1d2011e7d5d5ba26bea8bd4b0dec65f485f2b43/distributed/scheduler.py#L2611-L2613

off-topic but I spent some time on this before and it's worth sharing

you can rewrite this statement to

```python
pending: float = occ * (p - nc) / (p * nc)
pending = occ / nc - occ / p
pending = occupancy / n_threads - occupancy / processing
pending = avg_occ_per_thread - avg_occ_currently_processing
```

i.e. `pending` is "average occupancy per thread that is in state `ready` / not, yet executing"
or put into context of stealing, which is the only place that `saturated` matters, "occupancy per thread that could be stolen"
I guess what's written there is more stable in terms of floating point arithmetic.

So, the definition of saturated is spelled out something like

A worker is classified as saturated if it's `ready` occupancy is at least 0.4s per thread and at least 1.9 times the average cluster wide occupancy.

i.e. "there is something queued up / in ready and it is about two times more than average"

Since saturated only affects work stealing, I think we should not change anything about this definition in this PR

---

> Overall, my question is "do we need to track the total occupancy of queued tasks, like we currently do for processing?"

re: adaptive
I actually think that this "hack" is a fine estimation for queued occupancy. Whenever there is something queued up, chance are high that all the tasks contributing to total_occupancy are root tasks so this average makes sense. I'm wondering how stable this metric is once we transition from a queued dominated phase to a just-in-time assignment but the adaptive logic should be robust to short term fluctuations so this is not high priority


re: stealing
as soon as we approach a stealing-relevant stage where there are actually ready tasks on the worker, we're in the edge case of `queued_occupancy -> 0` so I think we're good unless there are workloads where we can have strong overlap between root-ish / queued and ready scheduling (I'm sure that's possible, I'm just wondering how likely/rare this is)

Reply:
TLDR I think we can keep total_occupancy as is for now

Reply:
I just realized this will be a pretty bad adaptive target actually. `total_occupancy` doesn't include tasks until they're in `processing`—which `queued` tasks aren't. So even if you have thousands of queued tasks, the total occupancy will only reflect the currently-running tasks, and will never recommend scale-up.

I think we would need to track occupancy of queued tasks, or add them to `total_occupancy`.

I can do that in a follow-up PR; for now I'll use the simpler but basically equivalent metric of `len(queued) * unknown_task_duration`.

xref https://github.com/dask/distributed/issues/6962